xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiAction.java

```diff
@@ -944,4 +946,25 @@ private boolean redirectSpaceURLs(String action, XWikiURLFactory urlf, XWiki xwi
 
         return false;
     }
+
+    /**
+     * Answer to a request with a JSON content.
+     * @param context the current context of the request.
+     * @param status the status code to send back.
+     * @param json the JSON to serialize in the content of the answer.
+     * @throws XWikiException in case of error during the serialization of the JSON.
+     */
+    protected void answerJSON(XWikiContext context, int status, JSONObject json) throws XWikiException
```

Comment:
I'm never a big fan of exposing publicly 3rd party classes (XWikiAction is public).

You could simply take an `Object` (or a Map since it seems to be your only use case right now) and use Jackson to serialize it.

Reply:
I haven't followed at all the PR but note that anything added to XWikiAction is potentially a technical debt since the new API is the Resource Handler and thus any feature added to XWiki Action also need to be provided to this new API (otherwise it'll make the removal of XWikiAction harder, hence the technical debt).

Reply:
> You could simply take an Object (or a Map since it seems to be your only use case right now) and use Jackson to serialize it.

I will use an Object in the signature.

> any feature added to XWiki Action also need to be provided to this new API 

Good to know. I actually added this method in XWikiAction to avoid duplicating it in SaveAction / SaveAndContinueAction. I can add it too in AbstractResourceReferenceHandler, but for now it won't be used and I'm not sure it will ever be, depending how we move those actions...

Reply:
> Good to know. I actually added this method in XWikiAction to avoid duplicating it in SaveAction / SaveAndContinueAction. I can add it too in AbstractResourceReferenceHandler, but for now it won't be used and I'm not sure it will ever be, depending how we move those actions...

Yes we'll need to convert these actions but we don't know when we'll do it. If the new code can be re-used as is then it'll make it easier when we do the move.